File: sdk/src/main/java/com/github/ddnosh/arabbit/module/rxbus/RxBus.java
package com.github.ddnosh.arabbit.module.rxbus;

import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import com.trello.lifecycle2.android.lifecycle.AndroidLifecycle;
import com.trello.rxlifecycle3.LifecycleProvider;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

import io.reactivex.Observable;
import io.reactivex.ObservableEmitter;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

public class RxBus {

    private static volatile RxBus mDefaultInstance;
    private final Subject<Object> mBus;
    private final Map<Class<?>, Object> mStickyEventMap;

    private RxBus() {
        mBus = PublishSubject.create();
        mStickyEventMap = new ConcurrentHashMap<>();
    }

    public static RxBus getDefault() {
        if (mDefaultInstance == null) {
            synchronized (RxBus.class) {
                if (mDefaultInstance == null) {
                    mDefaultInstance = new RxBus();
                }
            }
        }
        return mDefaultInstance;
    }

    /**
     * 发送事件
     */
    public void post(Object event) {
        mBus.onNext(event);
    }

    /**
     * 根据传递的 eventType 类型返回特定类型(eventType)的 被观察者
     */
    public <T> LifecycleObservable<T> toObservable(Class<T> eventType) {
        return new LifecycleObservable<>(mBus.ofType(eventType));
    }

    /**
     * 判断是否有订阅者
     */
    public boolean hasObservers() {
        return mBus.hasObservers();
    }

    /**
     * 发送一个新Sticky事件
     */
    public void postSticky(Object event) {
        synchronized (mStickyEventMap) {
            mStickyEventMap.put(event.getClass(), event);
        }
        post(event);
    }

    /**
     * 根据传递的 eventType 类型返回特定类型(eventType)的 被观察者
     */
    public <T> LifecycleObservable<T> toObservableSticky(final Class<T> eventType) {
        synchronized (mStickyEventMap) {
            LifecycleObservable<T> observable = toObservable(eventType);
            final Object event = mStickyEventMap.get(eventType);

            if (event != null) {
                Observable<T> source = observable.mergeWith(Observable.create(new ObservableOnSubscribe<T>() {
                    @Override
                    public void subscribe(ObservableEmitter<T> e) throws Exception {
                        e.onNext(eventType.cast(event));
                    }
                }));
                return new LifecycleObservable<>(source);
            } else {
                return observable;
            }
        }
    }

    /**
     * 根据eventType获取Sticky事件
     */
    public <T> T getStickyEvent(Class<T> eventType) {
        synchronized (mStickyEventMap) {
            return eventType.cast(mStickyEventMap.get(eventType));
        }
    }

    /**
     * 移除指定eventType的Sticky事件
     */
    public <T> T removeStickyEvent(Class<T> eventType) {
        synchronized (mStickyEventMap) {
            return eventType.cast(mStickyEventMap.remove(eventType));
        }
    }

    /**
     * 移除所有的Sticky事件
     */
    public void removeAllStickyEvents() {
        synchronized (mStickyEventMap) {
            mStickyEventMap.clear();
        }
    }

    public static class LifecycleObservable<T> extends Observable<T> {
        private Observable<T> source;

        public LifecycleObservable(Observable<T> source) {
            this.source = source;
        }

        public LifecycleObservable<T> bindLifeOwner(LifecycleOwner owner) {
            LifecycleProvider<Lifecycle.Event> provider = AndroidLifecycle.createLifecycleProvider(owner);
            return new LifecycleObservable<T>(compose(provider.bindToLifecycle()));
        }

        public LifecycleObservable<T> bindLifeProvider(LifecycleProvider<?> provider) {
            return new LifecycleObservable<T>(compose(provider.bindToLifecycle()));
        }

        @Override
        protected void subscribeActual(Observer<? super T> observer) {
            source.subscribe(observer);
        }
    }
}

File: sdk/src/main/java/com/github/ddnosh/arabbit/module/rxbus/RxBus.kt
package com.github.ddnosh.arabbit.module.rxbus

import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleEventObserver
import androidx.lifecycle.LifecycleOwner
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Consumer
import io.reactivex.subjects.PublishSubject

class RxBus private constructor() {

    companion object {
        private var bus: RxBus? = null
            get() {
                if (field == null) {
                    field = RxBus()
                }
                return field
            }

        fun get(): RxBus {
            return bus ?: RxBus()
        }
    }

    private val bus = PublishSubject.create<Any>()

    fun <T> on(
        eventType: Class<T>,
        owner: LifecycleOwner,
        onNext: Consumer<T>
    ) {
        val disposable = toObservable(eventType)
            .subscribe(onNext)
        toLifeCycle(owner, disposable)
    }

    fun <T> on(
        eventType: Class<T>,
        owner: LifecycleOwner,
        onNext: Consumer<T>,
        onError: Consumer<Throwable>
    ) {
        val disposable = toObservable(eventType).subscribe(onNext, onError)
        toLifeCycle(owner, disposable)
    }

    fun fire(event: Any) {
        bus.onNext(event)
    }

    private fun <T> toObservable(eventType: Class<T>?): Observable<T> {
        return bus.ofType(eventType)
    }

    private fun toLifeCycle(owner: LifecycleOwner, disposable: Disposable) {
        owner.lifecycle.addObserver(
            LifecycleEventObserver { _, event ->
                if (event == Lifecycle.Event.ON_DESTROY) {
                    disposable.dispose()
                }
            }
        )
    }
}
